docs/proposals/19-migrating-to-prometheus-operator.md.md

```diff
@@ -0,0 +1,444 @@
+gep-number: 0019
+creation-date: 2022-06-21
+status: implementable
+authors:
+- "@wyb1"
+- "@istvanballok"
+reviewers:
+- ""
+---
+
+# GEP-19: Monitoring Stack - Migrating to the prometheus-operator
+
+## Table of Contents
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Alternatives](#alternatives)
+
+## Summary
+
+As Gardener has grown, the monitoring configuration has also evolved with it.
+Many components must be monitored and the configuration for these components
+must also be managed. This poses a challenge because the configuration is
+distributed across the Gardener project among different folders and even
+different repositories (for example extensions). While it is not possible to
+centralize the configuration, it is possible to improve the developer experience
+and improve the general stability of the monitoring. This can be done by
+introducing the [prometheus-operator]. This operator will make it easier for
+monitoring configuration to be discovered and picked up with the use of the
+[Custom Resources][prom-crds] provided by the prometheus-operator. These
+resources can also be directly referenced in Go and be deployed by their
+respective components, instead of creating `.yaml` files in Go, or templating
+charts. With the addition of the prometheus-operator it should also be easier to
+add new features, such as Thanos.
+
+## Motivation
+
+Simplify monitoring changes and extensions with the use of the
+[prometheus-operator]. The current extension contract is described
+[here][extension-contract]. This document aims to define a new contract.
+
+Make it easier to add new monitoring features and make new changes. For example,
+when using the [prometheus-operator] components can bring their own monitoring
+configuration and specify exactly how they should be monitored without needing
+to add this configuration directly into Prometheus.
+
+The prometheus-operator handles validation of monitoring configuration. It will
+be more difficult to give prometheus invalid config.
+
+### Goals
+
+- Migrate from the current monitoring stack to the prometheus-operator.
+
+- Improve the monitoring extensibility and improve developer experience when
+  adding or editing configuration. This includes the monitoring extensions in
+  addition to core Gardener components.
+
+- Provide a clear direction on how monitoring resources should be managed.
+  Currently, there are many ways that monitoring configuration is being deployed
+  and this should be unified.
+
+- Leverage community knowledge by using [kube-prometheus] and
+  [kubernetes-mixins]. Reuse existing dashboards and monitoring configuration
+  for Kubernetes components whenever possible.
+
+- Improve how dashboards are discovered and provisioned for Grafana. Currently,
+  all dashboards are appended into a single configmap. This can be an issue if
+  ever exceed the maximum configmap size of 1MiB.
```

Comment:
Currently, grafana dashboards are even stored in a duplicated manner (extension configmap + aggregated configmaps that are actually mounted into grafana pods).
If you want, you could add to this goal as well, that we want to improve the scalability of this aspect by minimizing duplication of data (~> network transfer)

Reply:
That is a good idea! In addition to that since all of the dashboards are the same AFAIK, we could just generate the dashboards once in the `garden` namespace or another namespace and each grafana will discover these dashboards. If there an any exceptions to this, for example if a shoot with a certain feature active and has a specific dashboard, then this dashboard can be discovered via a special label set.

Reply:
That's an excellent idea, I like it!
Then let's allow both:
- placing common (non-specific) dashboards in a central place which is collected by all sidecars
- placing specific dashboards in the shoot namespace which is only collected by the respective sidecar (I can imagine k8s version-dependent dashboards, e.g. CSI)